### ISSUE TITLE
Inconsistency in the application of multiple recipe rules on the same field

### ISSUE THREAD

**Issue by wetzelj** [pydicom/deid#140]
After the parser update, there is an inconsistency in the manner in which multiple recipe rules applied to the same field.

Prior to the update, multiple rules were additive.  
**Example 1:**
```
JITTER StudyDate 1
JITTER StudyDate 2
```
Results in 3 days being added to StudyDate.

**Example 2:**
```
ADD PatientIdentityRemoved Yes
REMOVE PatientIdentityRemoved 
```
Results in no change to the header.

**Example 3:**
```
REMOVE StudyDate 
ADD StudyDate 20200806
```
Resulted in StudyDate having a value of 20200806.

After the parser update, any actions following a REMOVE rule are ignored. As a result, while Examples 1 and 2 still operate the same, Example 3 fails to add the new StudyDate to the header. 

I believe that this is occurring because the REMOVE action (specifically[ parser.py - delete_field()](https://github.com/pydicom/deid/blob/72cd3f6bd13865651521d2ac8481a9d09535a5a5/deid/dicom/parser.py#L158)) removes the specified field from the FileDataSet in the parser.dicom property but is not removing the item from the fields property.  In subsequent ADD rules, we fall into the add_field function and since the deleted field is still in parser.fields the add is treated as a replacement ([line 377](https://github.com/pydicom/deid/blob/72cd3f6bd13865651521d2ac8481a9d09535a5a5/deid/dicom/parser.py#L377)) which updates the element, but the tag with the new value is never re-added to the parser.dicom FileDataSet.

I have a few unit tests written for the examples that I've outlined above - but haven't implemented a solution yet (the new tests can be seen in this [commit](https://github.com/wetzelj/deid/commit/861e92f677538d4bc714d33c12eada3fd66de2c1)).  I wanted to get your perspective first.  My first thought was to ensure that the REMOVE action removed the element from the fields property in addition to what it currently does. 

While all of the above may seem like nonsense - "Why wouldn't you just compress each of these scenarios down to three single rules?" - the REMOVE/ADD scenario does have real-world applicability for situations like:
```
REMOVE endswith:ID
ADD PatientID 12345
```

**Comment by vsoch** [comment]
I think you are spot on - we'd want the field to not be found, and then have it [added here](https://github.com/pydicom/deid/blob/72cd3f6bd13865651521d2ac8481a9d09535a5a5/deid/dicom/parser.py#L386) so it does seem like a bug. So see if you are able to remove the field also from self.fields to address the issue! And very good debugging, as always. :)

**Comment by wetzelj** [comment]
Before I submit the pull request, could you take a look at [this commit](https://github.com/wetzelj/deid/commit/1fb38d5658aa6cf1cfbef76ad1830958c849ab79#diff-776cfd5c9387335de41894a2ea56de5f)? 

The removal of the value from fields introduced a bug detected by the [test_remove_all_func](https://github.com/pydicom/deid/blob/72cd3f6bd13865651521d2ac8481a9d09535a5a5/deid/tests/test_replace_identifiers.py#L567) unit test. The removal of the entry from fields caused a runtime error - "dictionary changed size during iterations".  To correct this, I created a clone of the fields dict which we then iterate over. Is there a better way to do this?

**Comment by vsoch** [comment]
That would work! I've typically seen that done more like:
```python
from copy import deepcopy
temp_fields = deepcopy(fields)
```
but if you are sure just doing dict(fields) would work (for all python versions) then I'm good with that. You could also do:

```python
for uid, field in fields.copy().items():
    ....
```
or even
```python
for uid in fields.keys():
    field = fields[uid]
    ...
```
As with Python, there are always many ways to do things! I usually do the first (with deepcopy) because I know some weirdness can result with deeply nested dictionaries.

**Comment by wetzelj** [comment]
I like typical. :) It's switched to deepcopy.

**Comment by vsoch** [comment]
Fixed with #141
